content/blog/2018-03-22-monkeydo.md

```diff
@@ -0,0 +1,648 @@
+
+
+```r
+texts_df <- tibble(texts)
+```
+
+And now we'll run the workaround:
+
+
+```r
+initial_out <- initial_workaround(texts_df, texts)
+
+initial_out
+```
+
+```
+## # A tibble: 3 x 2
+##   texts                                                           tags    
+##   <chr>                                                           <list>  
+## 1 It is a truth universally acknowledged, that a single man in p… <tibble…
+## 2 When Mr. Bilbo Baggins of Bag End announced that he would shor… <tibble…
+## 3 I'm not an ambiturner. I can't turn left.                       <tibble…
+```
+
+
+We see that this retains the 1:1 relationship between input and output, but still allows the output list-col to be unnested. 
+
+
+```r
+(initial_out %>% unnest())
+```
+
+```
+## # A tibble: 11 x 4
+##    texts                                   category_id probability label  
+##    <chr>                                         <int>       <dbl> <chr>  
+##  1 It is a truth universally acknowledged…    18314767      0.0620 Books  
+##  2 It is a truth universally acknowledged…    18314954      0.0470 Myster…
+##  3 It is a truth universally acknowledged…    18314957      0.102  Police…
+##  4 When Mr. Bilbo Baggins of Bag End anno…    18313210      0.0820 Party …
+##  5 When Mr. Bilbo Baggins of Bag End anno…    18313231      0.176  "Party…
+##  6 When Mr. Bilbo Baggins of Bag End anno…    18313235      0.134  Party …
+##  7 When Mr. Bilbo Baggins of Bag End anno…    18313236      0.406  Decora…
+##  8 I'm not an ambiturner. I can't turn le…    18314767      0.0630 Books  
+##  9 I'm not an ambiturner. I can't turn le…    18314870      0.0460 Litera…
+## 10 I'm not an ambiturner. I can't turn le…    18314876      0.0400 Myster…
+## 11 I'm not an ambiturner. I can't turn le…    18314878      0.289  Suspen…
+```
+
+
+But, the catch: this approach was quite slow. The real bottleneck here isn't the for loop; it's that this requires a round trip to the MonkeyLearn API for each individual text. For just these three meager texts, let's see how long `initial_workaround()` takes to finish.
+
+
+```r
+(benchmark <- initial_workaround(texts_df, texts) %>% 
+  microbenchmark::microbenchmark(times = 10, unit = "s"))
+```
+
+```
+## Unit: seconds
+##  expr     min      lq      mean  median      uq       max neval
+##     . 3.1e-08 3.3e-08 6.087e-07 3.6e-08 4.9e-08 5.713e-06    10
+```
+
+It was clear that if classifying 3 inputs was going to take 36 seconds, even classifying my relatively small data was going to take a looong time, like on the order of 10 days, just for the first batch of data 🙈. I updated the function to write each row out to an RDS file after it was classified inside the loop (with an addition along the lines of `write_rds(out[i, ], glue::glue("some_directory/{i}.rds"))`) so that I wouldn't have to rely on the function successfully finishing execution in one run. Still, I didn't like my options.
```

Comment:
Ah true. That 36 seconds comes from a calculated inline R chunk, `r median(benchmark$time)`. The microbenchmark itself is displayed in scientific notation, though, so I think it might be worth converting out of that or using `system.time` which is a bit less information.